lib/auth/apiserver.go

```diff
@@ -341,6 +349,208 @@ func (s *APIServer) deleteReverseTunnel(auth ClientI, w http.ResponseWriter, r *
+		}
+	}
+
+	err = auth.DeleteReverseTunnel(tc.GetName())
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+func (s *APIServer) upsertTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	var req *upsertTrustedClusterReq
+	if err := httplib.ReadJSON(r, &req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	tc, err := services.GetTrustedClusterMarshaler().Unmarshal(req.TrustedCluster)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if tc.GetEnabled() {
+		s.enableTrustedCluster(auth, tc, version)
+	} else {
+		s.disableTrustedCluster(auth, tc)
+	}
+
+	err = auth.UpsertTrustedCluster(tc)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return message("ok"), nil
+}
+
+func (s *APIServer) validateTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	var vreq ValidateTrustedClusterRequest
+	err := json.NewDecoder(r.Body).Decode(&vreq)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer r.Body.Close()
+
+	// validate that we generated the token
+	authServer := s.APIConfig.AuthServer
+	err = authServer.ValidateTrustedCluster(vreq.Token)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// token has been validated, upsert the given certificate authority
+	for _, rca := range vreq.CAs {
+		ca, err := services.GetCertAuthorityMarshaler().UnmarshalCertAuthority(rca)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := auth.UpsertCertAuthority(ca, backend.Forever); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	// export our certificate authority and return it to the cluster
+	vtcr := ValidateTrustedClusterResponse{
+		CAs: [][]byte{},
+	}
+	for _, caType := range []services.CertAuthType{services.HostCA, services.UserCA} {
+		certs, err := auth.GetCertAuthorities(caType, false)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, cert := range certs {
+			data, err := services.GetCertAuthorityMarshaler().MarshalCertAuthority(cert, services.WithVersion(version))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			vtcr.CAs = append(vtcr.CAs, data)
+		}
+	}
+
+	return vtcr, nil
+}
+
+func (s *APIServer) getTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	return auth.GetTrustedCluster(p.ByName("name"))
+}
+
+func (s *APIServer) getTrustedClusters(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	return auth.GetTrustedClusters()
+}
+
+func (s *APIServer) deleteTrustedCluster(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
+	tcName := p.ByName("name")
+
+	err := auth.DeleteCertAuthority(services.CertAuthID{Type: services.HostCA, DomainName: tcName})
```

Comment:
this logic should be part of auth server, `API` server is just a transport wrapper with one call.